### ISSUE TITLE
Add full support for installing location tags

### ISSUE THREAD

**Issue by mogoodrich** [mekomsolutions/openmrs-module-initializer#44]
Currently Iniz does not have full support for creating location tags outside of a location… currently, you just specific location tags by name “inline” in the locations.csv.

When saving a location, Iniz looks up each of these location tags by name, and creates them as needed.  

For this ticket, we want to make location tags first-class citizens.  

A first-pass at a design:

* Should support a locationtags.csv that allows specifying the location tag uuid, name and description

* The Locations loader should be updated so that it allows referring to tags by both uuid and name

* The Locations loader should also be updated so that if a “locationtags.csv” is present, it doesn’t “hot create” location tags as-needed as it currently does, rather if a location tag referenced in location.csv is not found, it should through a validation error; if no locationtags.csv file is present, it should continue to work as it currently does.

@mks-d thoughts on this?

fyi @mseaton  @brandones

**Comment by mseaton** [comment]
I 100% agree with making locationtags a first-class citizen, and for it to load ahead of the locations. Based on how other domains work (eg. Concept Class is a separate domain from Concept), it would seem that Location Tag should be a separate domain that is ordered ahead of Location.  This would be straightforward to implement.

Regarding how to handle this in relation to the current behavior with the "Tags" column in the Location domain, my vote would be to have this column retain it's current behavior for backwards compatibility.  Then, we would add support for additional columns in the same manner that Location Attributes are supported.  There would be columns with headers named (for example):

`Tag|Login Location`

With a column value that is either "true" or "false/blank".

This would only support looking up tags by name or uuid (based on the column value), it would not support hot-creating tags on demand.

This would lead to more intuitive Location CSVs, which - instead of having a bloated column value with lots of delimited tag names, simply has a bunch of extra columns with "true" or "" values indicating whether or not that Location has that tag.

This would also allow easier management of the metadata - one can easily take this CSV in Excel and sort it or filter it to show all of the Login Locations or all of the Visit Locations, etc.

**Comment by mks-d** [comment]
@mogoodrich 👍 sure, let's add that new domain.

@mseaton I like your idea, and obviously it is even better if you see Excel value for it.

The current "hot loading" was not properly thought through, and this is something that is overdue. Either to conclude that hot loading should be made smarter and more configurable, or should be deprecated (however, think of `ConceptReferenceTerm` for instance...). This is a discussion that can happen elsewhere. First of all I need to make an inventory of where exactly it happens, at worst in one or two other instances than location tags.

**Comment by mogoodrich** [comment]
Cool, thanks @mks-d ! So let's go with this then, with the caveat replacing my second two bullet points with the design that Mike lays out.

**Comment by mseaton** [comment]
@mks-d in addition to this enhancement, @mogoodrich made a good point recently that we also might want to provide the ability to link tags and locations in yet another separate CSV (eg. locationtagmap.csv).

The reason for this is that we are looking at having a "distribution-level" configuration (eg. for all of our Haiti sites, having one set of configurations), but needing to have a small number of configurations vary by implementation site.

In these cases, the locations and tags that we install will be the same across all of our Haiti locations, but the specific locations that are tagged as "login", "visit", etc. locations will be different at each facility implementation.

So having a locations.csv and locationtags.csv that are shared, and then a locationtagmap.csv that is custom per site might be the most straightforward way to implement this.

Thoughts or other design ideas?

**Comment by mks-d** [comment]
I think our approach would most likely be to override **locations.csv** for the children that differ from the base/reference config. But I can imagine that this might be too much hassle if there are multiple locations, multiple tags, _and_ many variations of their mappings.

Also this (obviously) introduces more decoupling, whereas Iniz should always try to enforce WYSIWYG when one looks at a CSV line for an entity. (I actually should write that on the main README.)

**Comment by mseaton** [comment]
Yes, but clearly the risk to that is that common location data grows out of sync.  If 95% of the location data is intended to be identical across facilities, with only the mappings to tags differing, then a solution that requires you to maintain 10 (or 50 or 1000) different versions of this same data to try to keep in sync and consistent is...suboptimal.

WYSIWYG is fine and great, but being a "Login Location" isn't an inherent attribute of a Location.  It is a separate table in the DB, and is more of a configuration setting for a specific server than a shareable attribute of a particular Location across a distribution footprint.

**Comment by mogoodrich** [comment]
Yes, to follow on what @mseaton , for instance in Haiti we have a several site-based servers, and the location list is identical between all of them, but the Thomonde server has the "Thomonde" location tagged as a Login Location, Consult Location, etc..., and then the St Marc location has the "St. Marc" location tags a Login Location, Consult Location, etc...  

We can (and do) handle this tag assignment outside of Iniz, but if we want to move this into Iniz, the locationtagmap.csv solution is the best solution I've thought of so far.

I do see however that this could be overly complex for implementations that don't have this complexity.

Would it be bad to support two configuration mechanism, both updated way of assigning location tags in the locations.csv that @mseaton suggest but also support the locationtagmaps.csv?  If so, should we keep this issue just about making Location Tags first-class citizens, and then open another issue to continue to debate the locationtagmaps.csv?

**Comment by mks-d** [comment]
Sure I get the use case, and I'm sure that if you brought it up it's because you have anticipated that challenge to come.

If one day Iniz is able to export all domains that it supports, and locations in particular, would you accept the tag mappings to be fed to the exported CSV for locations though?

**Comment by mks-d** [comment]
@mogoodrich 
>I do see however that this could be overly complex for implementations that don't have this complexity.

I thought @mseaton said that this was _in addition_ to the enhancement already described above in this thread? Otherwise yes indeed I think this should be another way to record those mappings, but not the only way.

**Comment by mseaton** [comment]
@mks-d I think that would be fine.  I'd only anticipate the export feature being used at the start of a given project, transitioning from non-iniz into iniz.  Do you think there would be a call to use this more on on ongoing basis?  Regarding your other comment, yes I think this is in addition to the enhancement already described.

**Comment by mks-d** [comment]
>Do you think there would be a call to use this more on on ongoing basis?

Yes, by running a proxy-OCL instance out of which metadata is periodically extracted. That's for ongoing and complex implementations phases.

But I agree that the primary use case is the one that you mentioned, i.e. transitioning to Iniz.

**Comment by brandones** [comment]
I think what's been laid out here is good. I like that the three approaches to Location Tags can all coexist. Tangentially, I really appreciate WYSIWYG being made a core value of Iniz.

I intend to implement this soon, assuming we're still happy with Mike's API suggestion for the Location, Location Tag, and Location Tag Mapping domains.

**Comment by brandones** [comment]
Actually, we still need to specify the Location Tag Mapping API. I'm having trouble imagining one that would be useful. @mseaton , what do you envision that CSV looking like?

**Comment by mogoodrich** [comment]
fwiw, in our design we will set up the Location Tag mappings via our configuration files (@brandones there's anoter ticket for that in the epic), so I don't think we need the Location Tag Mapping API (though I'm fine with adding it if we come up with a good design, but we can skip it if we get blocked on it)

**Comment by mseaton** [comment]
We might not, sorry.  I couldn't understand the question that was posed.  What did you mean by "I'm having trouble imagining one that would be useful"?  What are the outstanding questions?

**Comment by brandones** [comment]
All good. My question was, what would Location Tag Mapping CSVs look like?

**Comment by mseaton** [comment]
I'd imagine it would be as simple as possible:  a 2 column CSV, where the first column is the location name or uuid, and the second column is the tag name or uuid.

This basically matches exactly to the DB table.

**Comment by mogoodrich** [comment]
I think we are close! But @brandones just to make sure we are all on the same page, once you've figured out how you plan to proceed can you summarize here so we can confirm... I read through the above but it's not always 100% clear what decisions have been made.

And I agree with @mseaton re: location map API design.

Thanks!

**Comment by brandones** [comment]
1. To the locations domain, add support for tag columns like `Tag: Login Location` and `Tag: Visit Location`. These accept boolean values as interpreted by Iniz's [getBooleanFromKey](https://github.com/mekomsolutions/openmrs-module-initializer/blob/4798a9d351323034fe452c7ef7a4ddd18824c931/api/src/test/java/org/openmrs/module/initializer/api/InitializerServiceImplTest.java#L31). Tag columns for tags that don't exist cause a helpful ERROR and the CSV not to be loaded.
2. Add a location tag domain. It accepts CSVs with columns `UUID`, `Name`, `Description`, and `Void/Retire`.
3. Add a location tag map domain. It accepts CSVs with columns `Location Name`, `Location Tag Name`.

**Comment by mogoodrich** [comment]
Thanks @brandones , sounds good to me!

**Comment by brandones** [comment]
Uh-oh, I've hit a snag with the Location Tag Maps domain. How can Location Tags be removed from Locations? @mseaton @mogoodrich @mks-d ?

**Comment by mogoodrich** [comment]
@brandones do you mean how do you do it at the API level?

Check out setLocationTagsFor  methodn at LocationTagSetup:881 in PIH Core, specifically lines 894-900.

**Comment by brandones** [comment]
No, I mean at the user level. How do I tell Initializer that I want a location tag removed?

**Comment by brandones** [comment]
Should we add a boolean `Void/Retire` column that deletes the mapping? (Maybe we should call it `Remove`?)

**Comment by mogoodrich** [comment]
Ah, got it... I'd tend to defer to @mks-d  if there's any historical best practice we should use for this.

**Comment by brandones** [comment]
I've taken my best guess and called it "Void/Retire." PR is linked.

**Comment by mks-d** [comment]
Thanks a lot @brandones and @mogoodrich!
